doc/specs/stdlib_hash_functions.md

```diff
@@ -0,0 +1,1697 @@
+the tests and has no bad seeds.  To provide randomly generated seeds
+for the hash function the submodule also defines the subroutine
+`NEW_PENGY_HASH_SEED`. 
+
+The `stdlib_64_bit_spooky_hashes` submodule provides implementations
+of Bob Jenkins' SpookyHash in the form of the overloaded function,
+`SPOOKY_HASH`. Future implementations may provide the SpookyHash
+incremental hashing procedures.
+SpookyHash is optimized for large objects and should give excellent 
+performance for objects greater than about 96 byes, but has
+significant overhead for smaller objects.
+The code was designed for Little Endian processors, and will give
+different results on Big Endian processors, but the hash quality on
+those processors is probably just as good.
+SpookyHash version 2 passes all of Reini Urban's SMHasher tests, and
+has one bad seed only when reduced to a 32 bit output.
+Its only potential problem is undefined behavior if the key is
+misaligned.
+
+## The `stdlib_32_bit_hash_codes` module
+
+### Overview of the module
+
+Thirty two bit hash functions are primarily useful for generating hash
+codes for hash tables.
+Checksums generally benefit from having a larger number of bits.
+The `stdlib_32_bit_hash_codes` module defines five public overloaded
+32 bit hash code functions, `FNV_1`, `FNV-1A`, `NMHASH32`, `NMHASH32x`
+and `WATER_HASH`, two scalar hash functions, `FIBONACCI_HASH` and
+`UNIVERSAL_MULT_HASH`, four seed generators, `ODD_RANDOM_INTEGER` for
+`UNIVERSAL_MULT_HASH`, and `NEW_NMHASH32_SEED`, `NEW_NMHASH32X_SEED`,
+and `NEW_WATER_HASH_SEED`, for their respective hash code
+functions. It also defines the integer kind constant, `INT_HASH`, and
+a logical constant, `LITTLE_ENDIAN`, used to deal with one aspect of
+the machine dependence of the hash codes.
+
+### The `INT_HASH` parameter
+
+It is necessary to define the kind of integer used to return the hash
+code.
+As `stdlib_32_bit_hash_codes` deals exclusively with 32 bit hash codes,
+`INT_HASH` is an alias for the integer kind `INT32`.
+
+### The `LITTLE_ENDIAN` parameter
+
+In implementing hash functions it is sometimes necessary to know the
+"endianess" of the processor's integers. To this end the
+`stdlib_32_bit_hash_codes` module defines the logical parameter
+`LITTLE_ENDIAN` that, if true, indicates that the processor has little
+endian integers, and that if false indicates that the integers are big
+endian.
+
+### Specifications of the `stdlib_32_bit_hash_codes` procedures
+
+#### `FIBONACCI_HASH` - maps an integer to a smaller number of bits
+
+##### Status
+
+Experimental
+
+##### Description
+
+Calculates an `nbits` hash code from a 32 bit integer.
+
+##### Syntax
+
+`code = [[stdlib_32_bit_hash_codes:fibonacci_hash]]( key, nbits )`
+
+##### Class
+
+Pure function
+
+##### Arguments
+
+`key`: Shall be a scalar integer expression of kind `INT32`. It is an
+`intent(in)` argument.
+
+`nbits` Shall be a scalar default integer expression with `0 < nbits <
```

Comment:
So, does it return only positive integers?

Reply:
The Fibonacci hash only returns non-negative integers.